### PR TITLE
Add support for env processor

### PR DESCRIPTION
**Pull request by VincentLanglet**
Close https://github.com/phpstan/phpstan-symfony/issues/199

### CODE REVIEW

**Comment by VincentLanglet** [comment]
> @VincentLanglet Sorry, I'm busy with PHPStan 1.0, I'll have a look at this after the release.

No problem. Well done for the 1.0 release. 
I'm currently using 0.12 with my fork of the plugin to use this feature and the https://github.com/phpstan/phpstan-doctrine/pull/218 one. But I'll be happy to try the new 1.0 version.

If you have some time to take a look =)

Thanks

**Comment by ondrejmirtes** [comment]
@VincentLanglet  My top priority right now is PHP 8.1 support (https://phpstan.org/blog/plan-to-support-php-8-1). The main action is now happening in https://github.com/ondrejmirtes/BetterReflection/tree/ng and https://github.com/phpstan/phpstan-src/tree/ng-better-reflection. Once this is finished and released, the operations will return to normal eventually. I now have 187 emails in my inbox worthy of my attention (all related to PHPStan), so please stay tuned.

**Comment by VincentLanglet** [comment]
Friendly ping @ondrejmirtes,

Happy new year, I know you were  pretty busy with the support of php 8.1. I might be wrong but I understood you now finished the support. How long is your todo list now before having some time to take a new look to this PR and https://github.com/phpstan/phpstan-doctrine/pull/218 ?

Thanks a lot for all the work you do.

**Comment by ondrejmirtes** [comment]
Support for PHP 8.1 is not finished. I now have 213 unread emails in my inbox worth of my attention + still a lot of work to be done on PHP 8.1, like readonly properties, and updated php-8-stubs.

You don't need to ping me, when I get to it, I'll get to it.

File: src/Type/Symfony/ParameterDynamicReturnTypeExtension.php [hunk 1]
```diff
@@ -101,13 +115,45 @@ private function getGetTypeFromMethodCall(
...
+				)
+			);
+		}
+
+		if (
+			is_string($value)
+			&& preg_match('/%env\((.*)\:.*\)%/U', $value, $matches) === 1
+			&& strlen($matches[0]) === strlen($value)
+		) {
+			$providedTypes = EnvVarProcessor::getProvidedTypes();
```
Comment (lookyman) [inline src/Type/Symfony/ParameterDynamicReturnTypeExtension.php:149]: This introduces a hard dependency on `symfony/dependency-injection`. I get the idea about not duplicating Symfony's internal code here, but if we want to do it like that, we need to use something like `call_user_func(['\Symfony\Component\DependencyInjection\EnvVarProcessor', 'getProvidedTypes'])`.

Comment (VincentLanglet) [reply src/Type/Symfony/ParameterDynamicReturnTypeExtension.php:149]: I got the opposite review here: https://github.com/phpstan/phpstan-symfony/pull/200#discussion_r724885819

So I'll go with the class_exists solution.

Does 
```
if (
			is_string($value)
			&& preg_match('/%env\((.*)\:.*\)%/U', $value, $matches) === 1
			&& strlen($matches[0]) === strlen($value)
			&& class_exists(EnvVarProcessor::class)
		) {
			$providedTypes = EnvVarProcessor::getProvidedTypes();

			return $this->typeStringResolver->resolve($providedTypes[$matches[1]] ?? 'bool|int|float|string|array');
		}
```
wouldn't be enough ?

**Comment by VincentLanglet** [comment]
I added the class_exists check @lookyman 

(And fixed the conflicts)

**Comment by VincentLanglet** [comment]
> Support for PHP 8.1 is not finished. I now have 213 unread emails in my inbox worth of my attention + still a lot of work to be done on PHP 8.1, like readonly properties, and updated php-8-stubs.

I saw [this repository got some recent activity](https://github.com/phpstan/phpstan-symfony/releases) back and phpstan 1.4 is released with readonly property support. I'd like to avoid having to update my fork on every new release. Is there something I can do to help this PR to move forward @ondrejmirtes ? Do you need help on something else ?

**Comment by ondrejmirtes** [comment]
@lookyman has to give a 👍 and I’ll be happy to merge this.

**Comment by lookyman** [comment]
I will take a final look later this week. I'm sorry, I am currently swamped with other stuff.

**Comment by Wirone** [comment]
@VincentLanglet I have this errors and I found your MR - thank you very much! Before it's merged I would like to ask if it'll support such scenarios like here:

```yaml
parameters:
  some.bool: '%env(bool:FOO)%'
  is.still.bool: '%some.bool%'

services:
  Foo\Bar:
    public: true
    arguments:
      $isBool: '%some.bool%'
```

- will `$container->getParameter('is.still.bool')` be considered as `bool`?
- will `$isBool` in `Foo\Bar`'s constructor be considered as `bool`?

In general: will type resolved by `EnvVarProcessor` be propagated to "nested" parameters? We have large legacy app and I'm working on introducing this extension but our DI definition and container usage is... not optimal 😉

**Comment by Wirone** [comment]
Of course it's not a blocker, I just wanted to take the opportunity to ask before it's finished 🙂 Thanks for quick answer! Looking forward to get this merged but in the meantime I'll try to use fork and check it out.

**Comment by lookyman** [comment]
@ondrejmirtes Let's merge this!

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by Wirone** [comment]
I've installed 1.1.4 and just wanted to inform you that all places that previously was reported as errors (where parameters were accessed with `$container->getParameter()`) now are OK 🥳 Could remove one inline ignore and 2 "fixes" that I did before. Thank you very much once again @VincentLanglet 🙂

**Pull request merged**